vm/actor/src/builtin/multisig/mod.rs

```diff
@@ -146,36 +201,44 @@ impl Actor {
         rt.validate_immediate_caller_type(CALLER_TYPES_SIGNABLE.iter())?;
         let caller_addr: Address = *rt.message().caller();
 
-        let st: State = rt.state()?;
-        Self::validate_signer(rt, &st, &caller_addr)?;
+        rt.transaction(|st: &mut State, rt| {
+            if !is_signer(rt, st, &caller_addr)? {
+                return Err(actor_error!(ErrForbidden; "{} is not a signer", caller_addr));
+            }
+
+            let mut ptx = make_map_with_root(&st.pending_txs, rt.store()).map_err(
+                |e| actor_error!(ErrIllegalState; "failed to load pending transactions: {}", e),
+            )?;
 
-        rt.transaction::<State, _, _>(|st, rt| {
             // Get transaction to cancel
-            let tx = st
-                .get_pending_transaction(rt.store(), params.id)
-                .map_err(|err| {
-                    ActorError::new(
-                        ExitCode::ErrNotFound,
-                        format!("Failed to get transaction for cancel: {}", err),
-                    )
-                })?;
+            let tx = get_pending_transaction(&ptx, params.id).map_err(
+                |err| actor_error!(ErrNotFound; "Failed to get transaction for cancel: {}", err),
+            )?;
 
             // Check to make sure transaction proposer is caller address
             if tx.approved.get(0) != Some(&caller_addr) {
-                return Err(ActorError::new(
-                    ExitCode::ErrForbidden,
-                    "Cannot cancel another signers transaction".to_owned(),
-                ));
+                return Err(
+                    actor_error!(ErrForbidden; "Cannot cancel another signers transaction"),
+                );
             }
 
-            // Remove transaction
-            if let Err(e) = st.delete_pending_transaction(rt.store(), params.id) {
-                return Err(ActorError::new(
-                    ExitCode::ErrIllegalState,
-                    format!("Failed to delete transaction for cancel: {}", e),
-                ));
+            let calculated_hash = compute_proposal_hash(&tx, rt.syscalls()).map_err(|e| {
+                actor_error!(ErrIllegalState;
+                    "failed to compute proposal hash for (tx: {:?}): {}", params.id, e)
+            })?;
+
+            if params.proposal_hash != calculated_hash {
```

Comment:
specs-actor also checks if proposal_hash is nil (im guessing that for us is if its 0 len?) and if it is, i think this check passes.

Reply:
nah the empty check is unnecessary because calculated_hash can never be nil. There is no case where that check changes anything.

Even in their code I'm quite sure the nil sanity check is irrelevant because their unmarshalling of those params initializes the proposal hash as bytes so it can't be nil.